Add new flang driver main() entrypoint

CarolineConcatto opened the pull request:
### Important: This patch requires changes in upstream clang to land in order to have the correct behaviour.
---

Note' : This a working in progress to have a flang driver for 18. More
changes will be needed to fully support the new compiler.

Note'': Lit regression tests are added along with related CMakeLists
infrastructure, this may be the first patch to do so.

This work implements the "driver" part of an RFC sent to cfe-dev in [1].

The new entrypoint adds a binary called bin/flang.
The old flang.sh is removed from the install to make way for this.

bin/flang is implemented in terms of clang::Driver, defaulting to
the `--driver-mode=flang` mode recently added to clang in [1]. This
ensures that the driver runs in flang mode regardless of the name of
the binary inferred from `argv[0]`.
"flang -fc1" will be implemented as part of bin/flang in later patches.

If compiled against an old libclangDriver which is not flang-mode
aware, the driver fails with:

    "unsupported argument 'flang' to option '--driver-mode='"

Regression tests are added in test/driver, which mirror those in the
clang/test/Driver directory. The primary difference is that the clang
tests run `clang --driver-mode=flang <input>`, whereas these tests
run `flang <input>`.

In order to add regression tests, the relevant CMakeLists
infrastructure is included.

The configuration for the tests are in lit.cfg files. These rely on the
presence of llvm-lit and FileCheck, as long as F18 is not a part of the
monorepo it needs requires -DLLVM_EXTERNAL_LIT to be defined at cmake
configuration time to point to /bin/llvm-lit from an LLVM build.

[1] "RFC: Adding a fortran mode to the clang driver for flang"
    http://lists.llvm.org/pipermail/cfe-dev/2019-June/062669.html
[2] [clang][driver] Add basic --driver-mode=flang support for fortran
    https://reviews.llvm.org/D63607

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

CarolineConcatto committed:
Add new flang driver main() entrypoint

Note' : This a working in progress to have a flang driver for 18. More
changes will be needed to fully support the new compiler.

Note'': Lit regression tests are added along with related CMakeLists
infrastructure, this may be the first patch to do so.

This work implements the "driver" part of an RFC sent to cfe-dev in [1].

A new entrypoint adds a binary called bin/flang.
The old flang.sh is removed from the install to make way for this.
bin/flang is implemented in terms of clang::Driver, defaulting to
the `--driver-mode=flang` mode recently added to clang in [1]. This
ensures that the driver runs in flang mode regardless of the name of
the binary inferred from `argv[0]`.
"flang -fc1" will be implemented as part of bin/flang in later patches.

If compiled against an old libclangDriver which is not flang-mode
aware, the driver fails with:
  "unsupported argument 'flang' to option '--driver-mode='"

Regression tests are added in test/driver, which mirror those in the
clang/test/Driver directory. The primary difference is that the clang
tests run `clang --driver-mode=flang <input>`, whereas these tests
run `flang <input>`.

In order to add regression tests, the relevant CMakeLists
infrastructure is included.

The configuration for the tests are in lit.cfg files. These rely on the
presence of llvm-lit and FileCheck, as long as F18 is not a part of the
monorepo it needs requires -DLLVM_EXTERNAL_LIT to be defined at cmake
configuration time to point to /bin/llvm-lit from an LLVM build.

[1] "RFC: Adding a fortran mode to the clang driver for flang"
    http://lists.llvm.org/pipermail/cfe-dev/2019-June/062669.html
[2] [clang][driver] Add basic --driver-mode=flang support for fortran
    https://reviews.llvm.org/D63607

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

DavidTruby commented on tools/flang/main.cpp:
```diff
@@ -0,0 +1,81 @@
+static clang::DiagnosticOptions *CreateAndPopulateDiagOpts(
+    llvm::ArrayRef<const char *> argv) {
+  auto *DiagOpts = new clang::DiagnosticOptions;
+  unsigned MissingArgIndex, MissingArgCount;
```
`unsigned MissingArgIndex{}, MissingArgCount{};` 
This avoids having to disable the warning above. I realise though that these two values are set to 0 inside ParseArgs straight away so this isn't actually a bug.

CarolineConcatto replied:
I believe that is done in the CMakeFileLists.txt by removing almost all -wno

RichBarton-Arm replied:
I agree with David here. If you can reasonably change the code to avoid a warning then we should do that rather than disabling the warning. Is there a good reason to not do this?

Surely as compiler writers we don't like people who ignore the nice warnings we have striven hard to implement?! :-)

CarolineConcatto replied:
I believe this is fixed!

kiranchandramohan commented on tools/flang/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+  clangDriver
+  clangSerialization
+  clangEdit
+  clangLex
```
Is Lex required for Flang?

CarolineConcatto replied:
Done

kiranchandramohan commented on tools/flang/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+  Option
+)
+
+target_link_directories(flang
```
This requires cmake 3.13?

peterwaller-arm replied:
Ah, my bad. This should be removed.

The right thing to do is to use `find_package(Clang REQUIRED CONFIG)`. I think this should live along the `find_package(LLVM ...)` which lives in `/CMakeLists.txt` One slightly tricky thing will be to get `find_package` to pick up the right clang, ideally by searching in the `${LLVM_DIR}/../clang`, or similar. I think this can in principle be controlled by a `HINTS` argument to `find_package`, but I couldn't make it do the right thing quickly.

Here's a link to [`find_package`](https://cmake.org/cmake/help/v3.9/command/find_package.html) for cmake 3.9, which is the minimum supported cmake version specified in the root CMakeLists.

CarolineConcatto replied:
Done

kiranchandramohan commented on tools/flang/main.cpp:
```diff
@@ -0,0 +1,81 @@
+  llvm::SmallVector<const char *, 256> argv(argv_, argv_ + argc_);
+  auto TargetAndMode =
+      clang::driver::ToolChain::getTargetAndModeFromProgramName(argv[0]);
+  //Sets the driver mode to flang in case there is no flang mode in
```
clang-format is suggesting a space between the slashes and the comment. Applicable to next line also.

CarolineConcatto replied:
Replaced this line.
Done!

kiranchandramohan reviewed (commented):
A few comments.

DavidTruby commented on tools/flang/main.cpp:
```diff
@@ -0,0 +1,81 @@
+      TheDriver.BuildCompilation(argv));
+  llvm::SmallVector<std::pair<int, const clang::driver::Command *>, 4>
+      FailingCommands;
+  int Res = 1;
```
Or just not create the variable at all? 
e.g. `return TheDriver.ExecuteCompilation(*C, FailingCommands);`

CarolineConcatto replied:
Done

peterwaller-arm commented on tools/flang/main.cpp:
```diff
@@ -0,0 +1,81 @@
+  auto TargetAndMode =
+      clang::driver::ToolChain::getTargetAndModeFromProgramName(argv[0]);
+  //Sets the driver mode to flang in case there is no flang mode in
+  //clang driver. It avoids to call clang in any other mode
```
I think this also allows removing `#include "clang/Driver/ToolChain.h"`.

CarolineConcatto replied:
Done

peterwaller-arm commented on tools/flang/main.cpp:
```diff
@@ -0,0 +1,81 @@
+#include <memory>
+#include <set>
+#include <stdlib.h>
+#include <system_error>
```
I was able to make this include list much shorter:

```
#include "clang/Basic/Diagnostic.h"
#include "clang/Basic/DiagnosticIDs.h"
#include "clang/Basic/DiagnosticOptions.h"
#include "clang/Driver/Compilation.h"
#include "clang/Driver/Driver.h"
#include "clang/Driver/ToolChain.h"

// NOTE: These need to be removed as when libFlangFrontend has landed.
#include "clang/Frontend/CompilerInvocation.h"
#include "clang/Frontend/TextDiagnosticPrinter.h"

#include "llvm/ADT/ArrayRef.h"
#include "llvm/ADT/IntrusiveRefCntPtr.h"
#include "llvm/Option/ArgList.h"
#include "llvm/Option/Option.h"
#include "llvm/Option/OptTable.h"
#include "llvm/Support/InitLLVM.h"
#include "llvm/Support/VirtualFileSystem.h"
```

CarolineConcatto replied:
Done!

DavidTruby commented on tools/flang/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+    COMPILE_FLAGS
+      -Wno-unused-parameter
+      -Wno-unused-variable
+      -Wno-maybe-uninitialized
```
Can we not `include_directories(... SYSTEM)` the clang headers so that we aren't tainted by their lack of warning hygene?

CarolineConcatto replied:
Done

peterwaller-arm commented on tools/f18/CMakeLists.txt:
```diff
@@ -71,13 +71,3 @@ add_custom_target(module_files ALL DEPENDS ${MODULE_FILES})
-)
-file(RENAME "${CMAKE_CURRENT_BINARY_DIR}/bin/flang.sh" "${CMAKE_CURRENT_BINARY_DIR}/bin/flang")
-
-install(PROGRAMS flang.sh DESTINATION bin RENAME flang)
```
Further to the comment about renaming it to flang-tmp, this should be reinstated.

CarolineConcatto replied:
Done!

peterwaller-arm commented on tools/flang/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+add_executable(flang
```
Please update this and related targets to `flang-tmp`.

CarolineConcatto replied:
Done!

peterwaller-arm reviewed (commented):
This was just discussed on a private call: we need to leave bin/flang.sh in place. In the meantime we can give this new entrypoint a name like `flang-tmp`.

peterwaller-arm commented on tools/flang/main.cpp:
```diff
@@ -0,0 +1,81 @@
+#include "clang/Basic/DiagnosticIDs.h"
```
File is missing a license header.

CarolineConcatto replied:
Done!

CarolineConcatto commented on test/driver/flang.F90:
```diff
@@ -0,0 +1,51 @@
+! Check that flang -fc1 is invoked when in --driver-mode=flang.
```
I believe that these files are without header on LLVM PR.

CarolineConcatto replied:
Leaving like this for the moment.

peterwaller-arm commented:
I tried to run the tests compiling against llvm-project master, but they didn't work for me. It can't find `FileCheck`. I included in my CMake configuration:
```
-DLLVM_EXTERNAL_LIT=/path/to/llvm/build/bin/llvm-lit
-DLLVM_TOOLS_BINARY_DIR=/path/to/llvm/build/bin
```

Using `cmake --trace-expand`, I can see that `LLVM_TOOLS_BINARY_DIR` is overriding my `-D` option in `AddLLVM` to be the `<install prefix>/bin`.

I think this addition should to include some minimal instructions about how to run the tests currently towards the end of `README.md`.

CarolineConcatto commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+
+# suffixes: A list of file extensions to treat as test files.
+#  .f, .FOR, .for, .f77, .f90 and .f95 are common fortran file suffixes. And .F, .fpp and .FPP
+config.suffixes = ['.f', '.FOR', '.for', '.f77', '.f90', '.f95', '.F', '.fpp', '.FPP']
```
Done

CarolineConcatto commented on tools/flang/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+  clangSerialization
+  clangEdit
+  clangLex
+  clangBasic
```
Done!

RichBarton-Arm commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+#
```
Reading further, it looks like you have created this initial lit.cfg file by copy/pasting from (old) flang and adapting, which is sensible. It looks like you are basing it off a very old version though. There was a lot of refactoring of lit files done a few years back and the cfg files are a lot shorter. I got as far as the inferFlang call before remembering so there are a load of comments running up to that point, but few after. I suggest taking another look at what a modern clang lit.cfg file does and basing F18's off of that rather than off of old flang's.

CarolineConcatto replied:
This file has being removed. We added another PR for regression tests

CarolineConcatto replied:
This file now belongs to PR:https://github.com/flang-compiler/f18/pull/861.

RichBarton-Arm commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+    # -win32 is not supported for non-x86 targets; use a default.
+    return 'i686-pc-win32'
+
+config.substitutions.append( ('%clang_cc1',
```
This is all unnecessary if you base off a newer clang lit.cfg.py, but even then, I think we should operate on the principle of putting only what we need in our Flang lit.cfg.py. Like why do we need an %ms_abi_triple substitution?

CarolineConcatto replied:
This file has being removed

RichBarton-Arm commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+if not re.match(r'.*-win32$', config.target_triple):
+    config.available_features.add('non-ms-sdk')
+
+# Not set on native PS4 environment.
```
Similar comment above about only including needed features, etc.

CarolineConcatto replied:
This file has being removed

CarolineConcatto replied:
https://github.com/flang-compiler/f18/pull/861

RichBarton-Arm commented on test/lit.site.cfg.in:
```diff
@@ -0,0 +1,51 @@
+
+import sys
+
+config.llvm_src_root = "@LLVM_SOURCE_DIR@"
```
Similar comment about paring this down to just what we need.

CarolineConcatto replied:
This file has being removed

RichBarton-Arm commented on tools/flang/main.cpp:
```diff
@@ -0,0 +1,78 @@
+
+int main(int argc_, const char **argv_) {
+
+  llvm::InitLLVM X(argc_, argv_);
```
Redundant?

CarolineConcatto replied:
No, it is not. 
One of the things it does is allows us to debug the compiler in case it crashes
https://github.com/llvm/llvm-project/blob/5ad67df988f12f38b50f720c0fc4f1fa3b94a311/llvm/include/llvm/Support/InitLLVM.h#L16-L30

RichBarton-Arm reviewed (commented):
The driver entrypoint code looks good, but I think the lit.cfg needs a bit more thought and the commit should be split in two parts.

Good on you for adding lit capability though. It will be nice to port the test suite ASAP and this takes us a step closer.

peterwaller-arm commented:
Note: The first iteration of `--driver-mode=fortran` has now landed in upstream clang: llvm/llvm-project@6bf55804924d5a1d902925ad080b1a2b57c5c75c.

CarolineConcatto committed:
Regression tests configuration for f18 repository

The configuration for the tests are in lit.* files. These rely on the
presence of llvm-lit and FileCheck, as long as F18 is not a part of the
monorepo it needs requires -DLLVM_EXTERNAL_LIT to be defined at cmake
configuration time to point to /bin/llvm-lit from an LLVM build.

This patch:
 1) Changes CMakeLists.txt to:
  * Add the new folder with the regression tests configurations
  * Add lit configuration in test-lit/CMakeLists.txt
 2) Adds the test-lit folder with the bare minimum configuration of
    lit.cfg.py and lit.site.cfg.py.in to make a simple version lit test to work

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

CarolineConcatto commented on tools/flang/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+add_executable(flang-tmp
```
I'll change the message. The problem is that we want also to maintain the compatibility for test. I need to change the message.

CarolineConcatto commented on test/driver/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2019, Arm Ltd.  All rights reserved.
```
cmake complains when I remove. Maybe is some configuration

CarolineConcatto replied:
Removed extra  CMakes

CarolineConcatto committed:
Add LLVM header rule and more prunning of lit files

F18 is being built out of three for LLVM, therefore it needs to set
some variables.

Build LLVM with this environment variable set:
  -DLLVM_INSTALL_UTILS=On

Build f18 with these two environment variables:
  -DLLVM_INSTALLATION_DIR=$LLVM_DIR/bin
  -DLLVM_EXTERNAL_LIT=$LLVM_DIR/bin/llvm-lit

Otherwhise PATH and LD_LIBRARY_PATH needs to be set before running
the regression tests

llvm-lit, FileCheck, count and not are LLVM tools that regression
tests need to work

The regression tests are inside test-lit.

Run the test in the installation directory of f18 by:
make check-all

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

CarolineConcatto committed:
Fix version_test.f90 test to check full version and nothing besides that.
Fix misspling words in CMakeLists.txt and remove unused files from test-lit

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

CarolineConcatto committed:
Replacing test-lit/unit by test-lit/driver

sscalpone commented on test-lit/lit.cfg.py:
```diff
@@ -0,0 +1,68 @@
+
+
+# suffixes: A list of file extensions to treat as test files.
+config.suffixes = ['.f', '.F', '.FOR', '.for', '.f77', '.f90', '.F90',
```
Does this list of suffixes need to match the list in lit.local.config?

sscalpone replied:
Should we have a set of tests that verify the suffix list?  Is there a way to check that a test file *isn't* recognized?

CarolineConcatto replied:
So this patch has being rebase to this one:
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto committed:
Fix CMakeLists.txt

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

CarolineConcatto commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+# excludes: A list of directories to exclude from the testsuite. The 'Inputs'
+# subdirectories contain auxiliary inputs for various tests in their parent
+# directories.
+config.excludes = ['Inputs', 'CMakeLists.txt', 'README.txt', 'LICENSE.txt', 'ngrep.sh']
```
This file has being removed

CarolineConcatto commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+# Tweak the PATH to include the tools dir and the scripts dir.
+if flang_obj_root is not None:
+    flang_tools_dir = getattr(config, 'flang_tools_dir', None)
+    #if not flang_tools_dir:
```
This file has being removed

CarolineConcatto replied:
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto commented on test/driver/lit.local.cfg:
```diff
@@ -0,0 +1 @@
+config.suffixes = ['.f90', '.F90']
```
This file has being removed

CarolineConcatto replied:
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto commented on test/CMakeLists.txt:
```diff
@@ -12,6 +12,21 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+configure_lit_site_cfg(
```
The regression tests were added to another folder, but all the regression tests configuration is in PR861:
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto replied:
Done. We have test-lit, created by https://github.com/flang-compiler/f18/pull/861

CarolineConcatto commented on README.md:
```diff
@@ -79,6 +79,15 @@ where `LLVM_INSTALLATION_DIR` is
+
+LLVM_EXTERNAL_LIT=<LLVM_INSTALLATION_DIR>/bin/llvm-lit
+LLVM_TOOLS_BINARY_DIR= <LLVM_INSTALLATION_DIR>/bin
+
```
I've made some changes

CarolineConcatto replied:
So, now the regression tests are in:
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto committed:
Fix version_test by not accepcting extras characters in the output text
This test needs to be changed if the output of f18 -V changes.
At the momemt the ouput of $ ./path-to-bin/{f18/flang} -V is:

f18 compiler (under development)

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

CarolineConcatto committed:
The test version-test.f90 can check the start of the line as empty
 CMakeLists.txt makes used of LLVM cmake varibales, therefore
INSTALLATION_DIR and LLVM_LIT_EXTERNAL are  not needed when
building f18.

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

CarolineConcatto commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+
+    return flang
+
+def inferClang(PATH):
```
This file has being removed

CarolineConcatto replied:
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+                                 config.environment.get('LD_LIBRARY_PATH','')))
+    config.environment['LD_LIBRARY_PATH'] = path
+
+## Propagate path to symbolizer for ASan/MSan.
```
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+    # Get the source and object roots.
+    llvm_src_root = lit.util.capture(['llvm-config', '--src-root']).strip()
+    llvm_obj_root = lit.util.capture(['llvm-config', '--obj-root']).strip()
+    flang_src_root = os.path.join(llvm_src_root, "tools", "flang")
```
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto commented on test/lit.cfg:
```diff
@@ -0,0 +1,480 @@
+
+# Discover the 'flang' and 'clangcc' to use.
+
+import os
```
https://github.com/flang-compiler/f18/pull/861

CarolineConcatto committed:
Removes clang/frontend dependency
Creates a flangfrontend library with bare minimal.
The frontend only has the creater and destructor of the class.
Diagnostic printer will be implemented in the future.

Signed-off-by: Caroline Concatto <caroline.concatto@arm.com>

CarolineConcatto commented:
We have a dependency with Clang Frontend that was removed with the latest patch.

RichBarton-Arm reviewed (changes requested):
A few comments on the code and also a few global observations:

First, this PR now depends on the lit configuration changes which are now committed (hooray!). That means we can rebase this branch on top of master so I can review just the flang driver changes.

Second, with this patch we get a new driver binary called "flang-tmp" which passes some minimal options through to a binary called "flang" which is supposed to understand an option called "-fc1". With this patch applied we would have a binary called "flang-tmp" calling the script called "flang" which wraps the binary called "f18" which does not understand the -fc1 option. I think the intention is that this "flang" binary that "flang-tmp" is supposed to be calling does not exist yet? If so then this patch is internally inconsistent. It would be better to have "flang-tmp" call an as-yet nonexistant binary with the "fc1" option. Subsequent patches in this project would add said binary with the right functionality. When we are ready to ditch the current "flang" wrapper and/or "f18" binary in favour of the new ones then we can rename them at that point. I'd also be happy with another proposal that achieves the same aims of being internally consistent whilst the code is being built up.

Thirdly, playing around with some of the normal options on this new "flang-tmp" driver binary I get some slightly odd outputs:
```
 > flang-tmp --help
 OVERVIEW: clang LLVM compiler
USAGE: flang-tmp [options] file...

OPTIONS:
  -###                    Print (but do not run) the commands to run for this compila
(snip) (all options that the Clang driver supports)
> flang-tmp --version
clang version 10.0.0 (https://github.com/llvm/llvm-project.git e34801c8e6df11b802f58b202f2ce6598734f040)
...
```
I know that this adds a minimal first cut at an entry point, but it seems bad that this minimal first cut doesn't identify itself correctly from the start. I think we should change this as part of the original patch. I believe the driver simply needs to call to SetVersionPrinter to set the version. There is likely to be a similar API for overriding the default help screen if we can think of a sensible minimal list of options to include in this screen at this point. If that is complicated then I think 0 options would be better than all the options, including ones we'll never likely want to support.

CarolineConcatto commented on lib/frontend/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+  clangDriver
+  )
+
+include_directories(${LLVM_DIR} ${LLVM_DIR}/../..)
```
We need to include the llvm lib directory. I have replaced that for link_directories(${LLVM_LIBRARY_DIR}).
But libclangBasic and lib clang driver will only be found if the CMake can point to its binary otherwise we fail to build.

CarolineConcatto commented:
When we created this patch we wanted it to be able to call clang --driver-mode=flang from f18 side. 
And we would have for a while 2 flangs. 
flang-tmp that in the future would call fc1 and 
flang that is what we have currently inside f18. 
This 2 should coexist for a while until we could replace flang by using only flang-tmp.  

We had this discussion on:
 reviewed on 25 Sep 2019
Collaborator
peterwaller-arm left a comment
This was just discussed on a private call: we need to leave bin/flang.sh in place. In the meantime we can give this new entrypoint a name like flang-tmp.

RichBarton-Arm commented:
> When we created this patch we wanted it to be able to call clang --driver-mode=flang from f18 side.
> And we would have for a while 2 flangs.
> flang-tmp that in the future would call fc1 and
> flang that is what we have currently inside f18.
> This 2 should coexist for a while until we could replace flang by using only flang-tmp.

I agree with all that and remember the discussion. My point is that whilst we have the two flangs, the new flang needs to be internally consistent with its own design. I don't think it is ever intended for the new flang-tmp to call to the old flang? That is what would happen with this patch committed. 

I think the intention is for the new flang-tmp to call something similar to the current f18 binary. Maybe this binary is still TODO or maybe it can be the f18 binary itself. I would expect flang-tmp to either:
a) call some binary that does not exist yet, with the -fc1 argument. Calling flang without -### would emit an error, which we could perhaps catch and report a nicer version of depending on how it looks.
b) call the current f18 binary (without -fc1 argument)

I mildly prefer the former as it gives us a clean separation between the two drivers while under development, but I am open to suggestions otherwise.

peterwaller-arm commented on lib/frontend/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+
+set_target_properties(flangFrontend
+  PROPERTIES
+  COMPILE_FLAGS -Wno-unused-parameter)
```
If I recall correctly, because flang's default -Werror/-Wall flags turn it on, and the code as taken from flang was firing the warning (and therefore an error). I agree it should be removed and the related warnings fixed as necessary.

tskeith replied:
I think there is also code in some llvm headers that trigger this warning, so we will probably have to turn it off eventually. I don't think it's very useful and we should turn it off for all of f18.

klausler replied:
I have found the option to be useful and would like to retain it if possible.

CarolineConcatto commented:
After me and @RichBarton-Arm talked we discover that my patch is calling flang because Flang.cpp in llvm-project(llvm-project/clang/lib/Driver/ToolChains/Flang.cpp).
 It  has a hard code "flang"  implemented in it. So it will always call flang.
const char* Exec = Args.MakeArgString(D.GetProgramPath("flang", TC))
instead of  something similar to this:
const char *Exec = D.getClangProgramPath();
 A solution will be to change things inside llvm-project.

CarolineConcatto committed:
Add reference to the fortran frontend by setting it in the entrypoint.
$
This patch sets flang frontend according to the name of the driver.

To test the entrypoint and keep the previous behavior of flang this patch adds
flang-tmp as a tool in lit configuration files and adds the binary location in
FLANG_TOOLS_DIR.

As consequence the tests were modified to look for flang-tmp. Moreover,
"--driver-mode=flang" was removed as the flag is sent by the entry point in
tools/flang/main.cpp.

Important:
The entry point uses the flag create by the PR https://reviews.llvm.org/D73951
to set fortran frontend.

CarolineConcatto committed:
Update the name of the frontend flag used by flang driver.

Flang driver in -driver-mode=flang needs to set the flag -fortran-fe to call
its flang-fc1. Otherwise, the driver will call the generic flang-fc1 in the
clang driver[1].

This patch again needs the approval of the patch in [1] to work properly.

[1]https://reviews.llvm.org/D7

CarolineConcatto commented:
In my implementation, I assumed that the flang driver will select its frontend without users intervention and it will have the same name as the driver. 
For instance,
flang -fortran-fe test -###
will print  flang -fc1 and not test -fc1
the flang driver completely ignores test and will set frontran-fe = flang. 
Is it ok to set fortran-fe in flang driver, or not? Like -driver-mode is fixed to flang.  Or should the command line be able to overwrite the flag?
If that is ok, there is a second question.
With the current implementation, the frontend will have the same name as the driver. I was trying to follow clang behaviour when clang calls clang -fc1. Is this an ok assumption? According to the RFC proposed by Peter Waller, it looks like this is alright.
If that is not ok, the further question will come.

[1]http://lists.llvm.org/pipermail/cfe-dev/2019-June/062669.html

CarolineConcatto commented:
I did some mistake with my repo when I tried to do git pull --rebase.

CarolineConcatto commented:
@peterwaller-arm , @RichBarton-Arm , @kiranchandramohan  and @DavidTruby 
Do you mind if I close this PR and create a new one for this work? 
Lots have happened since I did this PR. First I rebased to lit, later to master. It now depends on other PR in llvm-project[1].
I would like to have a clean PR again. But is up to you.
[1]b7c8b9d

RichBarton-Arm commented:
> 
> 
> @peterwaller-arm , @RichBarton-Arm , @kiranchandramohan and @DavidTruby
> Do you mind if I close this PR and create a new one for this work?
> Lots have happened since I did this PR. First I rebased to lit, later to master. It now depends on other PR in llvm-project[1].
> I would like to have a clean PR again. But is up to you.
> [1]b7c8b9d

Starting afresh does sound like the cleanest way forward. You can link back to this for the full review trial.

RichBarton-Arm commented:
> 
> 
> In my implementation, I assumed that the flang driver will select its frontend without users intervention and it will have the same name as the driver.
> For instance,
> flang -fortran-fe test -###
> will print flang -fc1 and not test -fc1
> the flang driver completely ignores test and will set frontran-fe = flang.
> Is it ok to set fortran-fe in flang driver, or not? Like -driver-mode is fixed to flang. Or should the command line be able to overwrite the flag?
> If that is ok, there is a second question.
> With the current implementation, the frontend will have the same name as the driver. I was trying to follow clang behaviour when clang calls clang -fc1. Is this an ok assumption? According to the RFC proposed by Peter Waller, it looks like this is alright.
> If that is not ok, the further question will come.
> 
> [1]http://lists.llvm.org/pipermail/cfe-dev/2019-June/062669.html

I think the behaviour you want is to be able to set the default flang-fe inside the driver code to be "flang-tmp" by calling Driver.GenericFortranFE = "flang-tmp" in F18's driver.

So you want `flang-tmp` (with no other options - i.e. the default) to call "flang-tmp" and `flang-tmp --fortran-de="blah"` to call "blah". Looking at the way you have implemented the default in https://reviews.llvm.org/D73951, it looks like you can't do this. If you follow David's suggestion there and have the default value of GenericFortranFE be handled by the constructor then calling GetProgramPath on the class variable from inside ConstructJob() then the F18 driver could set GenericFortranFE to "flang-tmp" before calling constructJob and it would all work out. This line would later be removed when we rename flang-tmp to flang.

CarolineConcatto commented:
So, my code has this line in main.cpp

CarolineConcatto commented:
The file f18/tools/flang/main.cpp has this line:

RichBarton-Arm commented:
> But should it be F18 like this?
> flang-tmp -fortran-fe=foo-> calls frontend foo
> flang-tmp -> call frontend flang-tmp

I think it should be this one because:
 * It mirrors the clang --driver-mode=flang intention, i.e. "call the Flang Fortran frontend from the driver" from the flang driver.
 * It still allows the user to do something different if they want to override the default, same as they can for clang --driver-mode=flang
 * It naturally supports renaming of the flang compiler binary

To make that work, I think you would need to either:
 1. Check that the user has not already set the --fortran-fe option before setting the default in the current code.
 2. (on the clang side) Add a setter method to Driver for setting the GenericFortranFE member of Driver then call that setter (on the flang side) to set the new default value to argv[0] before you call Driver.BuildCompilation.

I think I mildly prefer option 2 as it kinda fits the existing pattern for setTargetAndMode in that class, but don't hold that opinion strongly.
